src/networking.c

```diff
@@ -1602,29 +1602,81 @@ int _writeToClient(client *c, ssize_t *nwritten) {
             c->sentlen = 0;
         }
     } else {
-        clientReplyBlock *o = listNodeValue(listFirst(c->reply));
-        size_t objlen = o->used;
+        if (listLength(c->reply) > 1) {
+            struct iovec iov[MAX_IOV_SIZE_PER_EVENT];
+            int iovcnt = 0;
+            /* The first node of reply list might be incomplete from the last call,
+             * thus it needs to be calibrated to get the actual data address and length. */
+            size_t offset = c->sentlen;
+            listIter iter;
+            listNode *next;
+            clientReplyBlock *o;
+            listRewind(c->reply, &iter);
+            while ((next = listNext(&iter)) && iovcnt < MAX_IOV_SIZE_PER_EVENT) {
+                o = listNodeValue(next);
+                if (o->used == 0) { /* empty node, just release it. */
+                    c->reply_bytes -= o->size;
+                    listDelNode(c->reply, next);
+                    continue;
+                }
 
-        if (objlen == 0) {
-            c->reply_bytes -= o->size;
-            listDelNode(c->reply,listFirst(c->reply));
-            return C_OK;
-        }
+                iov[iovcnt].iov_base = o->buf + offset;
+                iov[iovcnt].iov_len = o->used - offset;
+                if (++iovcnt == 1) offset = 0;
+            }
+            if (iovcnt == 0) return C_OK;
+            *nwritten = connWritev(c->conn, iov, iovcnt);
+            if (*nwritten <= 0) return C_ERR;
 
-        *nwritten = connWrite(c->conn, o->buf + c->sentlen, objlen - c->sentlen);
-        if (*nwritten <= 0) return C_ERR;
-        c->sentlen += *nwritten;
+            ssize_t remaining = *nwritten - (ssize_t)iov[0].iov_len;
+            if (remaining >= 0) { /* it's safe to release the first node. */
+                listRewind(c->reply, &iter);
+                next = listNext(&iter);
+                o = listNodeValue(next);
+                c->reply_bytes -= o->size;
+                listDelNode(c->reply, next);
```

Comment:
It seems these codes can be removed.
The following `while (remaining)` contains this condition.

Reply:
It's for `remaining == 0`.

Reply:
@panjf2000 I've tried to optimize this code, and it feels like it works.
No need for `if else` anymore.
```c
            size_t remaining = *nwritten;
            listRewind(c->reply, &iter);
            while (remaining) {
                next = listNext(&iter);
                o = listNodeValue(next);
                if (remaining < (ssize_t)(o->used - c->sentlen)) {
                    c->sentlen += remaining;
                    break;
                }
                remaining -= (ssize_t)(o->used - c->sentlen);
                c->sentlen = 0;
                c->reply_bytes -= o->size;
                listDelNode(c->reply, next);
            }
```

Reply:
@sundb